graylog2-server/src/main/java/org/graylog2/plugin/Message.java

```diff
@@ -466,18 +455,22 @@ public boolean removeStream(Stream stream) {
         return ImmutableSet.copyOf(this.indexSets);
     }
 
+    @SuppressWarnings("unchecked")
     public List<String> getStreamIds() {
-        if (!hasField(FIELD_STREAMS)) {
-            return streams.stream().map(Stream::getId).collect(Collectors.toList());
-        }
+        List<String> streamField;
         try {
-            @SuppressWarnings("unchecked")
-            final List<String> streamIds = getFieldAs(List.class, FIELD_STREAMS);
-            return new ArrayList<>(streamIds);
+            streamField = getFieldAs(List.class, FIELD_STREAMS);
         } catch (ClassCastException e) {
             LOG.trace("Couldn't cast {} to List", FIELD_STREAMS, e);
-            return Collections.emptyList();
+            streamField = Collections.emptyList();
+        }
+
+        final Set<String> streamIds = streamField == null ? new HashSet<>(streams.size()) : new HashSet<>(streamField);
+        for (Stream stream : streams) {
+            streamIds.add(stream.getId());
         }
+
+        return new ArrayList<>(streamIds);
```

Comment:
How about changing the signature of `getStreamIds()` to return a `Set` or create a different method for this case? This would avoid the additional list creation.

I just want to be super careful here because this stuff is called for every message that's written into Elasticsearch and now we are creating one HashSet and one ArrayList for every call. Previously this was one ArrayList if the message had streams.